#18 Limiter dashboard feedback Implementation, #14 debug fix

byBlurr opened the pull request:
RPM lights flash yellow if the limiter is enabled. flashes on for half second, off for half second.

I have tested this by reverting `if (!(ir_EngineWarnings.getInt() & irsdk_pitSpeedLimiter))` but have not tested it by enabling the pit limiter in sim. I do not have access to my pc with iRacing installed. Please test that the if statement works when enabling/disabling the limiter in sim. If it doesn't, then OverlayDDU.h line 261 needs looking into. If there are any other issues, I can look into them.

byBlurr committed:
Implement flashing pit limiter #18

- Needs testing, no iracing install

byBlurr committed:
Improved flashing LEDs #18

- Moved limiter lights out of rpm for loop.
- Tested by reverting if statement, not tested with iRacing client open.

byBlurr committed:
Added lap number to standings #9

lespalt commented:
I think I played around with irsdk_pitSpeedLimiter back then trying to do something similar, but it didn't quite work the way I expected. Please let me know once someone has had the chance to test this and I'll be happy to review/merge.

lespalt commented on OverlayDDU.h:
```diff
@@ -257,26 +258,71 @@ class OverlayDDU : public Overlay
+                    D2D1_ELLIPSE l6 = { float2(r2ax(0.5f - ww / 2 + (6 + 0.5f) * ww / 8),r2ay(0.065f)), r2ax(0.007f), r2ax(0.007f) };
+                    D2D1_ELLIPSE l7 = { float2(r2ax(0.5f - ww / 2 + (7 + 0.5f) * ww / 8),r2ay(0.065f)), r2ax(0.007f), r2ax(0.007f) };
+
+                    int frames = 60;
```
Would be nice if this was time-based rather than counting frames. I think there's some "blinking" logic in other places in the code.

byBlurr replied:
Agreed, I will take a look

byBlurr replied:
This is the only thing left to do for this implementation now. I am unsure where blinking has been used elsewhere? I'll take another skim through.

lespalt replied:
I might be misremembering about blinking used elsewhere.  It was in there at some point for something, but I might have removed it.

byBlurr replied:
For now I have opted for using the current tick count to work out if the lights should be on or off. Only thing is that I had to use a magic number (16) because the tick count was jumping 16 for every render call, however on odd occasions it would jump 17. There's definitely better ways of doing it, but this illuminates the need for incrementing a variable every tick.

The main rpm lights are moved out of the pit limiter if statement scope now, so this shouldn't affect performance at all if not using the limiter. I have fully tested this now on a laptop with lowest possible graphics settings, if anything would have performance issues it is this laptop. Tested on both 30Hz and 60Hz. I believe this is ready for merge.

lespalt commented on OverlayDDU.h:
```diff
@@ -257,26 +258,71 @@ class OverlayDDU : public Overlay
+                        }
+                    }
+                }
+                else
```
Can't the logic in the else branch be rolled into the above loop? We really just need additional logic that figures out the color and whether to use DrawEllipse or FillEllipse, right?

byBlurr replied:
That is how I originally did it, but it created a small delay between each LED being drawn, creating almost a wave affect. I moved it outside of the loop to remove that small delay and now all ellipse are visually drawn together.

byBlurr replied:
I have just improved on this. I have put the pit limiter into its own for loop, not sure why I didn't do this in the first place. Feel free to test putting the limiter code into the main for loop, when I do it, I seem to get the pit limiter flashing right to left.

byBlurr committed:
Changed pit limiter lights #18

- Moved draw into its own for loop
- Changed to only draw outline on LEDs, allowing RPM to still be displayed on LEDs when limiter is enabled
- Gear indicator background flashes green in sync with LED outline

byBlurr committed:
Reverted if statements back after testing #18

byBlurr commented:
I also committed a small change adding a lap counter to the standings overlay. So this PR also covers issue #9

byBlurr commented:
I am having no luck with fixing my issue #22... I am unable to diagnose it and therefor unable to test any features. I will leave it as is for now for someone else to test before I make any further changes.

byBlurr committed:
Issue #14 fix

- No longer crashes in debug mode

byBlurr commented:
This pull request now also includes the commit 8a9221c which fixes issue #14 (thanks to them for fixing it but not creating a pull request).

I have tested the pit limiter now and everything is working as expected. Tested in 60Hz and 30Hz mode.

byBlurr committed:
Pit limiter feedback timing #18 #21

- Changed how timing works for pit limiter feedback
- Would prefer to not use as many magic numbers, but I am limited by inconsistency in tick numbers

byBlurr commented:
Should be good for this pr to be merged. Will close issues #18 #9 and #14

byBlurr committed:
Revert "Added lap number to standings #9"

This reverts commit cddd59cec96d2345256db6cb41239874423cb26e.

Reverted as this would create a conflict with the @frmjar fork.

byBlurr commented:
a4d3118 reverts laps on leaderboard, as this would create a conflict with the @frmjar fork.
